### ISSUE TITLE
HTTP2 Capable Client

### ISSUE THREAD

**Issue by nickbabcock** [dropwizard/dropwizard#1569]
This issue is to track an HTTP2 capable client for `dropwizard-client`. Discussion about this topic started in #1349, but it's best if there was a separate issue. An http2 capable client won't make it in the release 1.0.

Copying @arteam's [comment (Nov. 2015)](https://github.com/dropwizard/dropwizard/pull/1349#issuecomment-158766533):

> Yes, `JerseyClient` and Apache `HttpClient` don't support HTTP/2 currently.
> 
> I know 3 Java client HTTP/2 libraries:
> - Jetty [Http2Client](https://github.com/eclipse/jetty.project/tree/master/jetty-http2/http2-client)
> - [OkHttp](https://github.com/square/okhttp)
> - Netty [Client](https://github.com/netty/netty/tree/master/example/src/main/java/io/netty/example/http2/helloworld/client) 
> 
> I like the first one from Jetty, because:
> - It provides an asynchronous API; 
> - `h2` and `h2c` support; 
> - Has the same core as `jetty-http2-server`. It provides a good interopability with the server implementation. Despite that HTTP/2 is a standard, there could be some > idiosyncrasy .

Since then, looks like Apache's `HttpClient` is [planning a major upgrade](http://hc.apache.org/status.html) to start transitioning to HTTP2 ([HTTPCLIENT-1672](https://issues.apache.org/jira/browse/HTTPCLIENT-1672)). If we want to go this route we'll have to wait for the release, and then implement the appropriate metrics in Dropwizard Metrics, and finally update dropwizard client. Additionally we may have to wait for HTTP2 support in Jersey (no idea if this is the case).

Replacing the client in dropwizard client would mean this could not be done until 2.x, as I'm assuming the APIs and configurations would change.

This is not an urgent issue (http 1.0/1 will be around for much much longer), but I want it to be tracked since dropwizard 1.0 will include support for a http2 server -- it is only natural to wonder the status of an http2 client.

**Comment by arteam** [comment]
Just for info, I started a [side-project](https://github.com/arteam/dropwizard-http2-client) several month ago for integration Dropwizard and the Jetty HTTP2 client. The client can be configured from a Dropwizard config, supports metrics instrumentation, and managed by the Dropwizard lifecycle.
Unfortunately, there is no much activity in the project, but it should be in a pretty a good shape.
I have a couple of unit and integration-tests, which periodically ping the HTTP/2 test [page](https://http2.golang.org/reqinfo) and validate responses, so I have some confidence in the project.

**Comment by arteam** [comment]
FYI, I published version 0.1 to the JCenter repository. Try it out,  should be available at
```xml
<dependency>
    <groupId>com.github.arteam</groupId>
    <artifactId>dropwizard-http2-client</artifactId>
    <version>0.1</version>
</dependency>
```
